src/main/typescript/wcardinal/ui/d-tree.ts

```diff
@@ -3,19 +3,207 @@
+			});
+
+			//update active state of all shown item when a child item is clicked
+			this._content.on("select", (selectedPosition: number[]): void => {
+				this._selectedPosition = selectedPosition
+				let items = this._content.children as Array < DTreeItem >
+				items.forEach(item => {
+					item.updateActiveState(selectedPosition)
+				});
+			})
+		}
+
+		protected update() {
+			const content = this._content
+			let	items = content.children as Array < DTreeItem > ;
+			content.height = this._itemOptionsShowable.length * this._itemHeight
+			const contentY = content.position.y
+			const height = this.height
+			const itemIndexMappedStart = (0 - contentY) / this._itemHeight - 2 < 0 ?
+				0 :
+				Math.floor((0 - contentY) / this._itemHeight) - 2
+
+			const itemIndexMappedEnd = (height - contentY) / this._itemHeight + 2 < this._itemOptionsShowable.length - 1 ?
+				Math.floor((height - contentY) / this._itemHeight) + 2 :
+				this._itemOptionsShowable.length
+
+			//get items options are shown in Dpane content
+			const itemOptionsShown = this._itemOptionsShowable.slice(itemIndexMappedStart, itemIndexMappedEnd)
+
+			if (items.length < itemOptionsShown.length) {
+				for (let i = items.length; i < itemOptionsShown.length; i++) {
+					const itemOptions = itemOptionsShown[i]
+					const treeItem = new DTreeItem(itemOptions)
+					content.addChild(treeItem)
+					//listen select item event
+					treeItem.on(UtilPointerEvent.down, (): void => {
+						this.emit("select", treeItem)
+						if (treeItem.isParent()) {
+							this.onToggle(itemOptions.itemPosition)
+						}
+					});
+				}
+			} else if (items.length > itemOptionsShown.length) {
+				for (let i = itemOptionsShown.length; i < items.length; i++) {
+					content.removeChild(items[i])
+					items.length = itemOptionsShown.length
+				}
+			}
+			for (let i = 0; i < items.length; i++) {
+				items[i] = items[i].updateItem(itemOptionsShown[i], this._selectedPosition)
+			}
+		}
+
+		protected onToggle(itemPosition: number[]) {
+			if (this._value instanceof Array) {
+				let items = this._value
+				//access to data of selected item in this.value to update expanded attribute
+				for (let i = 0; i < itemPosition.length - 1; i++) {
+					items = items[itemPosition[i]].children
+				}
+				const item = items[itemPosition[itemPosition.length - 1]]
+				item.expanded = !item.expanded
+
+				//reset data of tree widget
+				this._itemOptions = {}
+				this._itemOptionsShowable.length = 0
+				this._itemY = 0
+				this._content.removeAllListeners()
+				this._content.removeChildren()
+
+				//re-render tree
+				this.mapped([], this._value)
+				this.update()
+				this.registerListeners()
+			}
+		}
+
+		private mapped(parentPosition: number[], items: Array<DTreeItemJsonData>) {
+				for (let i = 0; i < items.length; i++) {
+					const item = items[i];
+					if (item) {
+						//position of item in this._value
+						const itemPosition: number[] = parentPosition.concat([i])
```

Comment:
Please try to minimize the GC pressure.

Reply:
I don't really understand this comment.
Should I use "parentPosition.concat([i])" directly instead of creating and using "itemPosition"?

Reply:
In my application, the number of items could be tens of thousands at the worst case. When the number of items are that large, this method creates numbers of array instances even if the data (_itemOptions) exists. One of the options we have is reusing the _itemOptions if it exits.

Reply:
In my application, the number of items could be tens of thousands at the worst case. When the number of items are that large, this method creates numbers of array instances even if the data (_itemOptions) exists. One of the options we have is reusing the _itemOptions if it exits.

Reply:
I have reused "_itemOptions" and updated data in "_itemOptions" by "key" is itemPosition.join("-")
I'm sorry, I don't really understand your comments.
Could you give me some detail suggestions or examples?

Reply:
Something like the follwoing:
```
private mapped( parentPosition: number[], items: Array<DTreeItemJsonData> ) {
	for (let i = 0; i < items.length; i++) {
		const item = items[i];
		const key = itemPosition.join('-');
		let itemOptions : DTreeItemOptions | undefined = this._itemOptions[ key ];
		if( itemOptions != null ) {
				// Update the item options
		} else {
			itemOptions = { ... };
			this._itemOptions[ key ] = itemOptions;
		}
	}
}
```